### PR TITLE
Add support for local files to CsvFileSource

### LINKED ISSUE #2281: Support file system resources in @CsvFileSource
## Problem description

Test resources are strange beasts. The compiler does not know about them nor cares, so getting the resources into the class path in typical builds needs help from the build system (ant, maven, gradle). This leads to several problems

- running a single test file (e.g. a first after project clean or fresh project) will not work, unless a more or less complete build step is executed.
- copying the test file from one location to another without any transformation is a waste of time (and ssd wear).
- Modifying the test data and rerun the test requires a rebuild (remember the deployment of the test data to the (test) class path, so modifying just the test data and re-run the test is not enough.
- Test data should be considered input to the program, not a resource. A resource is a glorified constant, test data should be able to vary independent of program code, test or otherwise.

## Original Proposal
- Modify the InputStream creating method (see my remarks in #1970) to try file system first, then class path would be a sufficient change. This will create this feature very cheaply without changing current class path semantics.
- Update @CsvFileSource documentation to reflect this change since the change proposal in #1970 adapts a private method in the CsvFileArgumentsProvider class.

### PR DESCRIPTION
**Pull request by marcphilipp**
In addition to classpath resources, CsvFileSource now supports files
from the local file system.

Resolves #2281.

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/master/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileArgumentsProvider.java [hunk 1]
```diff
@@ -160,4 +161,50 @@ private void advance() {
...
+
+	}
+
+	private static class DefaultInputStreamProvider implements InputStreamProvider {
+
+		private static final DefaultInputStreamProvider INSTANCE = new DefaultInputStreamProvider();
+
+		@Override
+		public InputStream openClasspathResource(Class<?> baseClass, String path) {
+			Preconditions.notBlank(path, "Classpath resource [" + path + "] must not be null or blank");
```
Comment (sbrannen) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileArgumentsProvider.java:192]:
```suggestion
			Preconditions.notBlank(path, () -> "Classpath resource [" + path + "] must not be null or blank");
```

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileArgumentsProvider.java [hunk 2]
```diff
@@ -160,4 +161,50 @@ private void advance() {
...
+			InputStream inputStream = baseClass.getResourceAsStream(path);
+			return Preconditions.notNull(inputStream, () -> "Classpath resource [" + path + "] does not exist");
+		}
+
+		@Override
+		public InputStream openFile(String path) {
+			Preconditions.notBlank(path, "File [" + path + "] must not be null or blank");
```
Comment (sbrannen) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileArgumentsProvider.java:199]:
```suggestion
			Preconditions.notBlank(path, () -> "File [" + path + "] must not be null or blank");
```

File: junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileArgumentsProvider.java [hunk 3]
```diff
@@ -160,4 +161,50 @@ private void advance() {
...
+			try {
+				return Files.newInputStream(Paths.get(path));
+			}
+			catch (IOException e) {
+				throw new UncheckedIOException("File [" + path + "] could not be read", e);
```
Comment (sbrannen) [inline junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileArgumentsProvider.java:204]: Shouldn't this rather be a `JUnitException` (or subclass thereof)?

Comment (marcphilipp) [reply junit-jupiter-params/src/main/java/org/junit/jupiter/params/provider/CsvFileArgumentsProvider.java]: You're right. I changed it.

**Pull request merged**